---
title: "MES Timecourse JS"
author: "Jared Slosberg"
date: "5/5/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r init, results = 'hide', message = F}
#source('scripts/init.R')
library(BUSpaRse)
library(monocle3)
library(DropletUtils)
library(ggplot2)
library(org.Mm.eg.db)
library(dplyr)
library(biomaRt)
library(xtable)
library(scales)
library(scran)
library(here)
library(purrr)
library(here)

source(here("scripts/accessory_functions/doDifferentialExpression.R"))
source(here("scripts/accessory_functions/assignCC.R"))

directs <- c("data","plots","results")
map(directs, function(dir){
  if(!dir.exists(dir)){
    dir.create(here(dir))
  }
})

```

# Create monocle sc object from kallisto pipeline output
# Generate knee plots and filter each sample by a minimum UMI count per cell
```{r, cache = T}
samples<-c("TH","TL") 


cds_list<-list()
spliced_kallisto_path <- here("preprocessing/bus_output/")
# Read genes.mtx
for (i in 1:length(samples)){
    print(samples[i])
    mat<-t(Matrix::readMM(paste0(spliced_kallisto_path,samples[i],"/spliced/s.mtx")))
    gene_info<-readr::read_tsv(paste0(spliced_kallisto_path,samples[i],"/spliced/s.genes.txt"), col_names = FALSE)
    barcodes<-readr::read_tsv(paste0(spliced_kallisto_path,samples[i],"/spliced/s.barcodes.txt"), col_names = FALSE)
    
    cds<-new_cell_data_set(mat,
                         cell_metadata = barcodes,
                         gene_metadata = gene_info)
    
    #cds <- cds[,Matrix::colSums(exprs(cds)) != 0] 
    #cds <- estimate_size_factors(cds)
    pData(cds)$sample<-samples[i]
    pData(cds)$batch<- pData(cds)$sample
    
    #cds@int_metadata$UMI_cut <- UMI_cut_by_sample[i]
    
    cds_list<-c(cds_list,cds)
}

# Find low UMI cells
numExpected<-6000 #cells targeted
UMI_cutoff<-1000
lapply(1:length(samples), function(sampleIdx){
    print(samples[sampleIdx])
    bc_rank <- barcodeRanks(exprs(cds_list[[sampleIdx]]), lower = 80)
    
    print(qplot(bc_rank$total, bc_rank$rank, geom = "line") +
      geom_vline(xintercept = metadata(bc_rank)$knee, color = "blue", linetype = 2) +
      geom_vline(xintercept = metadata(bc_rank)$inflection, color = "green", linetype = 2) +
      geom_vline(xintercept = UMI_cutoff, color = "darkred", linetype = 2) +
      geom_hline(yintercept = numExpected, color = "black", linetype = 2) +
      annotate("text", y = c(1000,5000,10000), x = 1.5 * c(metadata(bc_rank)$knee, metadata(bc_rank)$inflection,UMI_cutoff),
               label = c(
                   paste0("knee= ",metadata(bc_rank)$knee," UMIs"), 
                   paste0("inflection= ",metadata(bc_rank)$inflection," UMIs"),
                   paste0("UMI_cutoff = ",UMI_cutoff)
                   ),
                   color = c("blue", "green","darkred")) +
      scale_x_log10() +
      scale_y_log10() +
      labs(y = "Barcode rank", x = "Total UMI count") + 
      ggtitle(paste0("Knee Plot - ",samples[sampleIdx]))
    )
})


#How many cells would remain for different UMI cut offs?
UMI_num_cells <- list("starting.droplets" = NA, "UMI.at.knee" = NA, "cells.at.knee" = NA, "cells.at.expected.UMI" =NA, "UMI.at.expected.cells" = NA)
UMI_num_cells <- lapply(1:length(samples), function(sampleIdx){
    print(samples[sampleIdx])
    #tot_counts<-Matrix::colSums(exprs(cds_list[[sampleIdx]]))
    bc_rank <- barcodeRanks(exprs(cds_list[[sampleIdx]]), lower = 80)
    
    
    UMI_num_cells$starting.droplets <- bc_rank@nrows
    UMI_num_cells$UMI.at.knee <- bc_rank@metadata$knee
    UMI_num_cells$cells.at.knee <- unique(bc_rank@listData$rank[as.character(bc_rank@listData$total) == as.character(bc_rank@metadata$knee)])
    UMI_num_cells$cells.at.expected.UMI <- unique(bc_rank@listData$rank[which(bc_rank@listData$total == UMI_cutoff)])
    UMI_num_cells$UMI.at.expected.cells <- min(bc_rank@listData$total[bc_rank@listData$rank < numExpected])
    
    return(UMI_num_cells)
})

```

Do the filtering at specified cutoffs
```{r}
# Filter emtpy droplets
UMI_cutoff<-200 
filtered_cds_list<-lapply(cds_list,function(cds){
  pData(cds)$Total_UMIs<-Matrix::colSums(exprs(cds))
  #bc_rank <- barcodeRanks(exprs(cds))
  #print(metadata(bc_rank)$knee)
  #print(metadata(bc_rank)$inflection) 
  filtered_cds<-cds[, pData(cds)$Total_UMIs > UMI_cutoff]
  return(filtered_cds)
})
 
#mitochondrial filtering is in a chunk below, after gene name conversion

# How many cells remain per sample?
lapply(filtered_cds_list,function(x){dim(x)[2]})

#Combine cds files
#cds<-combine_cds(filtered_cds_list, cell_names_unique = TRUE)
cds<-do.call(cbind,filtered_cds_list)

colnames(fData(cds))<-c("gene_id")
rownames(cds)<-fData(cds)$gene_id
colnames(pData(cds))[1]<-"barcode"
rownames(pData(cds))<-paste(pData(cds)$barcode,pData(cds)$sample,sep=".")
colnames(cds)<-rownames(pData(cds))

# Summary stats
cds<-detect_genes(cds, min_expr = 0)
fData(cds)$num_reads <- Matrix::rowSums(exprs(cds))

pData(cds)$log10UMI <- log10(pData(cds)$Total_UMIs)

```

## Get and add gene symbols from biomart
```{r}
#With BioMart, get gene names, make sure to use the right ensembl archive for genome used for alignment

ensembl <- useMart("ensembl", host = "http://jul2019.archive.ensembl.org/") #mm 97
datasets <- listDatasets(ensembl)
ensembl <- useDataset("mmusculus_gene_ensembl",mart=ensembl)

fData(cds)$gene_id_trimmed <- stringr::str_split_fixed(fData(cds)$gene_id,"\\.",2)[,1]
gene_description <- getBM(attributes = c("ensembl_gene_id_version","external_gene_name","description","chromosome_name"),
                   filters = "ensembl_gene_id", values = fData(cds)$gene_id_trimmed,  mart = ensembl)

#Prepare info as data frame for merging with sc object
gene_description[,"gene_id_trimmed"] <- stringr::str_split_fixed(gene_description[,1],"\\.",2)[,1]
gene_description <- gene_description[,c("gene_id_trimmed","external_gene_name")]

fData(cds) <- left_join(as.data.frame(fData(cds)), gene_description, by ="gene_id_trimmed")
colnames(fData(cds))[colnames(fData(cds)) == "external_gene_name"] <- "gene_short_name"

fData(cds)$gene_short_name <- make.unique(fData(cds)$gene_short_name)

```


# Now with gene names, calculate mitochondrial gene percentage in each cell for quality assessment. Filter on UMI and mt ratio
```{r}

mito_genes<-fData(cds)$gene_id[grepl("^mt-",fData(cds)$gene_short_name)]

pData(cds)$mt_reads <- Matrix::colSums(exprs(cds)[mito_genes,])
pData(cds)$total_reads  <- Matrix::colSums(exprs(cds))
pData(cds)$mito_ratio <- pData(cds)$mt_reads/pData(cds)$total_reads

#filter here on mito genes if desired
for(i in 1:length(samples)){
  print(samples[i])
  print(sum(pData(cds)$sample == samples[i]))
  }

cds <- cds[, pData(cds)$mito_ratio < .20]

use_genes <- Matrix::rowSums(exprs(cds)) > 0
cds <- cds[use_genes,]

for(i in 1:length(samples)){
  print(samples[i])
  print(sum(pData(cds)$sample == samples[i]))
}

```

# Preprocess merged CDS
```{r}
cds <- estimate_size_factors(cds)
cds <- preprocess_cds(cds, num_dim = 50, verbose=TRUE)


plot_pc_variance_explained(cds)

assay(cds, "logcounts") <- normalized_counts(cds, norm_method = "log")

```

# Dimensionality Reduction
```{r}
set.seed(041)

do.existing.embedding <- F
if(do.existing.embedding){
  pdat <- read.csv(here("results/lmmp_umap_clusters_man.csv"), row.names = 1)
  
  assertthat::assert_that(all(rownames(pdat) == colnames(cds)))
  
  reducedDim(cds,"UMAP") <- pdat[, c("UMAP_1","UMAP_2")]

  pData(cds)$clusters <- pdat[,"clusters"] %>% as.factor()
} else{

  cds <- reduce_dimension(cds,max_components=2,cores=6,verbose=TRUE)

}


plot_cells(cds)
plot_cells(cds,color_cells_by="sample",cell_size=0.5,label_cell_groups = F)

plot_cells(cds,genes=c("Sox10","Ret","Uchl1","Snap25"),cell_size=0.8)

#plot_cells(cds, genes=c("Met", "Cdh3", "Clic3", "Cftr", "Myh6", "Calcb"),cell_size=0.5)
#plot_cells(cds, genes=c("Ret", "Sox10", "Nos1", "Snap25", "Gfra1", "Gfra2"),cell_size=0.5)
#plot_cells(cds, genes=c("Slpi", "Wt1", "Msln", "Uchl1", "Chat", "Gpr88", "Best1", "Krt19", "Gfap", "Nes"),cell_size=0.5)

```


# Remove batch effect
```{r}
#do not run if using existing embedding, which is loaded in block above
if(!do.existing.embedding){
  set.seed(502)
  
  cds <- align_cds(cds, num_dim = 50, residual_model_formula_str = "~batch + num_genes_expressed",alignment_k = 20,verbose=T)
  cds <- reduce_dimension(cds,max_components=2,cores=4,verbose=TRUE)
  #cds <- order_cells(cds)
  
  plot_cells(cds,color_cells_by="sample",cell_size=0.5, label_cell_groups = F, alpha = .5)
  plot_cells(cds, genes = c("Ret","Sox10"), alpha = 0.5)

}
```

# Clustering 
```{r}

if(!do.existing.embedding){
  cds <- cluster_cells(cds, k=15, cluster_method="leiden",num_iter=5, reduction_method="UMAP", partition_qval=0.05,verbose=TRUE,resolution=1e-05)
  pData(cds)$clusters<-clusters(cds)
  
} 

plot_cells(cds, color_cells_by = "clusters")
plot_cells(cds, color_cells_by="sample") + ggtitle("Sample")



p<-ggplot(as.data.frame(pData(cds))) +
  geom_bar(aes(x=clusters,fill=sample),position="fill") + scale_color_brewer() + monocle3:::monocle_theme_opts() + ggtitle("sample representation by cluster")

#pdf("~/Goff_Lab/Kulkarni_ENS/6month_LMMP/plots/Cluster_by_sample.pdf",width=9,height=6)
plot_cells(cds) 
#plot_cells(cds, color_cells_by = "sample", label_cell_groups = F) + scale_color_brewer()
plot_cells(cds, color_cells_by = "log10UMI") 

#dev.off()

```

# Marker Gene testing
```{r, message= F, results= 'hide', warning=F}
marker_test_res <- top_markers(cds, group_cells_by="clusters", cores=1,verbose=F, genes_to_test_per_group = 30)

write.table(x = marker_test_res,file = here("results/cluster_markers.tsv"),sep = "\t",quote = F, col.names = T)
```

```{r}
top_specific_markers <- marker_test_res %>%
                            filter(fraction_expressing >= 0.10) %>%
                            group_by(cell_group) %>%
                            top_n(2, pseudo_R2)

top_specific_marker_ids <- unique(top_specific_markers %>% pull(gene_id))

plot_genes_by_group(cds,
                    top_specific_marker_ids,
                    group_cells_by="clusters",
                    ordering_type="maximal_on_diag",
                    max.size=3)

top_specific_markers <- marker_test_res %>%
                            filter(fraction_expressing >= 0.00) %>%
                            group_by(cell_group) %>%
                            top_n(30, pseudo_R2)

dt <- DT::datatable(top_specific_markers)

top_specific_markers <- top_specific_markers[order(top_specific_markers$cell_group),]


```

# QC plots
```{r}
p<-ggplot(as.data.frame(pData(cds))) 

p1<-p + geom_violin(aes(x=sample,y=log10(Total_UMIs),fill=sample)) + monocle3:::monocle_theme_opts() + ggtitle("Total UMIs by sample")+ guides(fill=FALSE)

p2<-p + geom_violin(aes(x=sample,y=log10(num_genes_expressed),fill=sample)) + monocle3:::monocle_theme_opts() + ggtitle("Num Genes Expressed by sample") + guides(fill=FALSE)

p4 <- p +
      geom_violin(aes(x=clusters, y=log10(Total_UMIs), fill = clusters)) +
      labs(x= "clusters", y ="log10 UMI count") +
      scale_color_brewer(palette = "Set1") +
      theme(legend.position = "none") +
      ggtitle("cluster vs UMI count") +
      monocle3:::monocle_theme_opts()

p5 <- p +
      geom_violin(aes(x=clusters, y=log10(num_genes_expressed), fill = clusters)) +
      labs(x= "clusters", y ="log10 genes expressed") +
      scale_color_brewer(palette = "Set1") +
      theme(legend.position = "none") +
      ggtitle("cluster vs genes expressed") +
      monocle3:::monocle_theme_opts()

p1
p2

p4
p5

```

# Gene expression plots for annotations
```{r, eval = T}
genes_of_interest <- c("Smo","Fmo2","Aebp1","Cftr","Clic3","Slpi","Met","Cdh3","Calcb")

pdf(file = here("plots/MENS_markers.pdf"), width = 20, height = 20)
plot_cells(cds, genes = genes_of_interest, alpha = 0.5, group_label_size = 4)
dev.off()
```

# Annotations (manual)
```{r, eval =T}

#These clusters and annotations are based on the loaded embedding
if(do.existing.embedding){

  pData(cds)$cell_type<-"Unknown"
  
  pData(cds)$cell_type[pData(cds)$clusters == 1] <- "Macrophage-A"
  pData(cds)$cell_type[pData(cds)$clusters == 2] <- "MENs"
  pData(cds)$cell_type[pData(cds)$clusters == 3] <- "Macrophage-B" 
  pData(cds)$cell_type[pData(cds)$clusters == 4] <- "Neuroglia"
  pData(cds)$cell_type[pData(cds)$clusters == 5] <- "Smooth muscle cells" #smtn, Acta2
  pData(cds)$cell_type[pData(cds)$clusters == 6] <- "Vascular endothelium"
  pData(cds)$cell_type[pData(cds)$clusters == 7] <- "RBC"
  pData(cds)$cell_type[pData(cds)$clusters == 8] <- "B Lymphocytes" #
  pData(cds)$cell_type[pData(cds)$clusters == 9] <- "Pdgfra+ Fibroblasts" #
  pData(cds)$cell_type[pData(cds)$clusters == 10] <- "Penk+ Fibroblasts" #
  pData(cds)$cell_type[pData(cds)$clusters == 11] <- "T cells" #
  pData(cds)$cell_type[pData(cds)$clusters == 12] <- "NK cells" #
  
  pData(cds)$cell_type[pData(cds)$clusters == 13] <- "NENs" #snap25, epha5, 
  pData(cds)$cell_type[pData(cds)$clusters == 14] <- "Macrophage-C" 
  pData(cds)$cell_type[pData(cds)$clusters == 15] <- "Interstitium" 
} else{
  
  
  pData(cds)$cell_type<-"Unknown"
  
  pData(cds)$cell_type[pData(cds)$clusters == 1] <- "MENs"
  pData(cds)$cell_type[pData(cds)$clusters == 2] <- "Macrophage-A"
  pData(cds)$cell_type[pData(cds)$clusters == 3] <- "Macrophage-B" 
  pData(cds)$cell_type[pData(cds)$clusters == 4] <- "Neuroglia"
  pData(cds)$cell_type[pData(cds)$clusters == 5] <- "Vascular endothelium"
  pData(cds)$cell_type[pData(cds)$clusters == 6] <- "Smooth muscle cells" #smtn, Acta2
  pData(cds)$cell_type[pData(cds)$clusters == 7] <- "RBC"
  pData(cds)$cell_type[pData(cds)$clusters == 8] <- "RBC"
  pData(cds)$cell_type[pData(cds)$clusters == 9] <- "B Lymphocytes" #
  pData(cds)$cell_type[pData(cds)$clusters == 10] <-  "Pdgfra+ Fibroblasts" #
  pData(cds)$cell_type[pData(cds)$clusters == 11] <- "Penk+ Fibroblasts"   
  pData(cds)$cell_type[pData(cds)$clusters == 12] <- "Macrophage-G2M"
  pData(cds)$cell_type[pData(cds)$clusters == 13] <- "T cells"
  pData(cds)$cell_type[pData(cds)$clusters == 14] <- "NENs"
  pData(cds)$cell_type[pData(cds)$clusters == 15] <- "NK cells"
  pData(cds)$cell_type[pData(cds)$clusters == 16] <- "NK cells"
  pData(cds)$cell_type[pData(cds)$clusters == 17] <- "Smooth muscle cells B"
  pData(cds)$cell_type[pData(cds)$clusters == 18] <- "Interstitium"
  pData(cds)$cell_type[pData(cds)$clusters == 19] <- "Macrophage-C" 
  pData(cds)$cell_type[pData(cds)$clusters == 20] <- "NENs" 
  
}
```

```{r do_diff_expression}
lrt <- doDifferentialExpression(cds_obj = cds, method = "LRT",min_pct_cells_exprs =  .02, 
                                full = "~num_genes_expressed + mito_ratio + clusters",
                                reduced = "~num_genes_expressed + mito_ratio")

write.csv(lrt, file = here("results/de_genes_by_cluster.csv"))
lrt_filt <- lrt %>% filter(q_value < 0.05)
print(qplot(lrt_filt$num_cells_expressed))

#ensembl ids character vector
de_genes <- pull(lrt_filt, gene_id)

```

```{r save}
saveRDS(cds, file = here("6month_LMMP.rds"))
write.csv(colnames(cds)[pData(cds)$cell_type == "MENS"], here("results/MENS/6mo_mens_names.txt"),quote = F, row.names = F)
```

```{r}
if(!file.exists(here("results/lmmp_umap_clusters_man.csv"))){

  pdat <- cbind(as.data.frame(reducedDim(cds, "UMAP")), as.data.frame(pData(cds)$clusters), as.data.frame(reducedDim(cds, "Aligned")))
  colnames(pdat) <- c("UMAP_1","UMAP_2","clusters",paste0("Aligned_PC",1:50))

}
```

# Session Information
```{r session}
sessionInfo()
```

